storage/innobase/lock/lock0lock.cc

```diff
@@ -1052,21 +1076,17 @@ static void lock_wait_wsrep(trx_t *trx)
          back and the lock conflict would be resolved when an XA COMMIT
          or XA ROLLBACK statement is executed in some other connection.
 
-         If victim has also BF status, but has earlier seqno, we have to wait.
+         If BF trx doesn't have to wait for victim trx, then BF abort
+         is possible.
       */
       if (lock->trx != trx && lock->trx->mysql_thd &&
-          !(wsrep_thd_is_BF(lock->trx->mysql_thd, false) &&
-            wsrep_thd_order_before(lock->trx->mysql_thd, trx->mysql_thd)))
+          !wsrep_BF_has_to_wait(lock, trx))
       {
-        if (wsrep_thd_is_BF(lock->trx->mysql_thd, false))
+        // There is no need to kill BF victim with compatible lock.
+        if (wsrep_thd_is_BF(lock->trx->mysql_thd, false) &&
+            !lock_has_to_wait(trx->lock.wait_lock, lock))
         {
-          // There is no need to kill victim with compatible lock
-          if (!lock_has_to_wait(trx->lock.wait_lock, lock))
-            continue;
-
-#ifdef UNIV_DEBUG
-          wsrep_report_error(lock, trx);
```

Comment:
Well, if we bring semantics of wsrep transactions in InnoDB, there's no way to simply report BF-BF error for every such a case, because now it may be BF waiting on SR to complete or in prepared state, or BF waiter to be running SR by itself, thus error output produced by wsrep_report_error() becomes false-positive.

Although the current fix is also incorrect, I will need to think more of it.
The problem right now is how to merge wsrep transaction checks with lock_has_to_wait() checks, and also report proper BF-BF error cases.